doc/TockBinaryFormat.md

```diff
@@ -0,0 +1,127 @@
+# Tock Binary Format
+
+<!-- npm i -g markdown-toc; markdown-toc -i Compilation.md -->
+
+<!-- toc -->
+
+- [TBF Header](#tbf-header)
+  * [TLV Types](#tlv-types)
+    + [`1` Main](#1-main)
+  * [`2` Writeable Flash Region](#2-writeable-flash-region)
+  * [`3` Package Name](#3-package-name)
+- [Code](#code)
+
+<!-- tocstop -->
+
+Tock processes are represented using the Tock Binary Format (TBF). A TBF
+includes a header portion, which encodes meta-data about the process, followed
+by a binary blob which is executed directly.
+
+## TBF Header 
+
+The TBF header contains a base header, followed by a sequence of
+type-length-value encoded elements. The base header 16 bytes, and has 5 fields:
+
+```
+ 0         2             4            8       12         16
++---------+-------------+------------+-------+----------+
+| Version | Header Size | Total Size | Flags | Checksum |
++---------+-------------+------------+-------+----------+
+```
+
+    * `Version` a 16-bit unsigned integer specifying the TBF header version.
+      Always `2`.
+    * `Header Size` a 16-bit unsigned integer specifying the length of the TBF
+      header in bytes.
+    * `Total Size` a 32-bit unsigned integer specifying the total size of the TBF
+      in bytes (including the header).
+    * `Flags` each bit indicates whether a flag is enabled (1) or disabled (0).
+      - Bit 0 marks the process enabled. Disabed processes will not be launched
+        at startup.
+    * `Checksum` the result of XORing each 4-byte word in the header, excluding
+      the word containing the checksum field itself.
+
+The header is followed immediately by a sequence of TLV elememnts.  Each
+element begins with a 16-bit type and 16-bit length followed by the element
+data:
+
+```
+ 0      2        4
++------+--------+-----...---+
+| Type | Length | Data      |
++------+--------+-----...---+
+```
+
+  * `Type` is a 16-bit unsigned integer specifying the element type.
+  * `Length` is a 16-bit unsigned integer specifying the size of the data field
+    in bytes
+  * `Data` is the element specific data. The format for the `data` field is
+    determined by its `type`.
+
+### TLV Types
+
+TBF may contain arbitrary element types. A standard set of element types are
+standardized.
+
+#### `1` Main
+
+The `Main` element has three 32-bit fields:
```

Comment:
Note that `Main` is required for processes and that without a main TLV the kernel will treat the "process" as padding and skip to the next TBF header in flash.

Reply:
Three options:

1. OK

2. Keep it required, but do we actually need `protected_size`? Don't we _always_ want to protect the entire header?

3. Remove `protected_size` for the same reason, but also `Main` doesn't have to be required, we'd just use default values of `min_ram_size: 0` and `init_offset: [first-byte-after-the-header]`.

Reply:
Yes we always want to protect the header, but the goal is to allow the system to protect more space. So if the kernel wants to store nonvolatile state for the process it has somewhere to put it. We could, I suppose, create a "kernel state" TLV (or something) and make it part of the header and then just say that the header is protected.

I still think we want a way to have empty apps for padding in the linked list of apps in flash. We could use a flag in the flags field.

Reply:
> Yes we always want to protect the header, but the goal is to allow the system to protect more space

OK, so what if we keep `protected_size`, but it means protected bytes after the header? And defaults to `0` if `Main` is not present?

> I still think we want a way to have empty apps for padding in the linked list of apps in flash. We could use a flag in the flags field.

Sure. Isn't this already covered by the `enabled` flag? I.e. a disabled app will already be treated as just padding.

Reply:
> OK, so what if we keep protected_size, but it means protected bytes after the header? And defaults to 0 if Main is not present?

That's fine, but process.rs and elf2tbf will have to be updated to reflect that.

>Sure. Isn't this already covered by the enabled flag? I.e. a disabled app will already be treated as just padding.

That puts tockloader in a little bit of a bind. It wouldn't really know how to show legitimate disabled apps and hide just padding, and it wouldn't know how to prevent users from enabling just padding.

Reply:
> That's fine, but process.rs and elf2tbf will have to be updated to reflect that.

Will do

> That puts tockloader in a little bit of a bind. [...]

Hrm... I see what you're saying. Wouldn't it be better to use a more intrinsic property of the header. For example, header version `0` is not a valid version (and since we're currently on `2` it never will be), so you could mark padding that way. `tockloader` could use an intentionally incorrect checksum or header size (e.g. 0)...

In any case for now, I don't even think it's bad for `tockloader` to require the `Main` TLV header, and we can still avoid making it required for Tock to load and run a process correctly.

Reply:
I've update three things:

1. I made `Main` optional in `process.rs`, but never elided in elf2tbf, so tockloader can rely on it being there, but it's not required to actually run a process.

2. I changed the semantics of `init_fn_offset` and `protected_size` to both be offsets from the binary payload (e.g. after the header)

Reply:
I think this is an OK compromise, but we can't just use an arbitrarily invalid header as padding because that's also how we denote the end of the list of apps. But I think that making a header that is completely valid per this spec, has no TLVs, but has version 0 instead of 2 will work.